### PR TITLE
Evaluation tool

### PR DESCRIPTION
**Pull request by KirillTim**
While working on https://github.com/marco-c/crashsimilarity/issues/39 I faced many issues and decided to start big refactoring.
Add cache(100MB in ram for all archived signatures) to speedup signature access operations.
Add downloads cache to reduce network operations time.
Introduce abstract `Algorithm` class and start working on slow reference implementation of WMDistanceModel, which will be base line for all future models
Comments are welcome :)

### CODE REVIEW

**Comment by KirillTim** [comment]
@ibrahimsharaf I will try not to change anything related to you work in existing code to simplify future merge. But you can have a look at my classes hierarchy to make you algorithm compatible with mine

File: refactored/cache.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    def __init__(self, traces=None, downloader_cache=None):
+        # TODO: construct from Dict[signature, List[(stack_trace, uuid)] ?
+        self.downloader_cache = downloader_cache if downloader_cache else dict()
+        self.traces = traces if traces else []
+
+    @staticmethod
+    def build(stream):
+        """build from downloaded archives"""
+
+        # TODO: is it possible to have different `signature`s for one `proto_signature` ?
```
Comment (marco-c) [inline refactored/cache.py]: No, there should be only one signature for each proto_signature (that is basically a stack trace).

**Comment by marco-c** [comment]
Overall, it looks good so far.

Is the cache actually needed? How slow is it to rebuild it? If it isn't slow, we could remove the cache altogether (it's better to avoid premature optimizations).

**Comment by KirillTim** [comment]
@marco-c Yes, the cache is actually needed. It dramatically speedup experiments evaluating.
> How slow is it to rebuild it?
It cost no time to rebuilt the cache, while it is built during reading. The only problem is disk space. Model+corpus take ~600mb of disk space

**Comment by KirillTim** [comment]
Evaluation tool seems to be done. 
We also have baseline solution based on cosine distance now. IMO, https://github.com/marco-c/crashsimilarity/issues/29 can be closed now.
https://github.com/marco-c/crashsimilarity/issues/11 can be closed as well.
In future we will need to extract evaluation into function to simplify progress measuring.

**Comment by marco-c** [comment]
I really like the changes, but could you split them in separate PRs?

One PR to refactor the code with the classes; one PR to refactor the downloader; one PR to introduce the cache; one PR to add the basic evaluation tool.

> It is not reasonable to write unit tests for testing model accuracy, but I can write some parts of evaluation tool logic as unit tests. Should I add it here or on the new PR, after this will be merged(I prefer the second option)?

Yes, this can be an additional and separate PR.

**Comment by marco-c** [comment]
> @marco-c Yes, the cache is actually needed. It dramatically speedup experiments evaluating and network operations.

The only problem is that for the evaluation we will use periods of time far from each other, so the cache might not help much in those cases.

**Comment by KirillTim** [comment]
> The only problem is that for the evaluation we will use periods of time far from each other, so the cache might not help much in those cases.

Yes, for sure. Cache won't help much in production, and will be turned off. But while we are in stage of active development it will help us :)

**Comment by KirillTim** [comment]
Last part of this PR is implemented in https://github.com/marco-c/crashsimilarity/pull/88

**Pull request closed without merging**